File: Batch4ProblemSolving/merge-intervals.java
//https://leetcode.com/problems/merge-intervals

class Solution {
    public int[][] merge(int[][] intervals) {

        //step: sort intervals with start time 
        Arrays.sort(intervals, new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                return a[0] - b[0];
            }
        });
        
        //step: iterate over intervals, merge overlapping with end time >= start
        LinkedList<int[]> mergedList = new LinkedList<>();
       
        for(int[] current: intervals) {
            if(mergedList.isEmpty() || current[0] > mergedList.getLast()[1]) {
                mergedList.add(current);
            } else {
                mergedList.getLast()[1] = Math.max(mergedList.getLast()[1], current[1]);
            }
        }
        
        return mergedList.toArray(new int[mergedList.size()][]);
    }
}

File: Batch4ProblemSolving/merge-intervals.kt

//https://leetcode.com/problems/merge-intervals/
class Solution {
    fun merge(intervals: Array<IntArray>): Array<IntArray> {
        //step: sort intervals by start time
        // intervals.sortWith(object: Comparator<IntArray> {
        //     override fun compare(a: IntArray, b: IntArray): Int {
        //         return a[0] - b[0]
        //     }
        // })
        
        intervals.sortWith(Comparator{ a, b -> a[0] - b[0] })
        
        //merged to non overlaping list
        var merged: MutableList<IntArray> = mutableListOf()
        
        for(current in intervals) {
            if(merged.isEmpty() || current[0] > merged[merged.lastIndex][1]) {
                merged.add(current)
            } else {
                merged[merged.lastIndex][1] = maxOf(merged[merged.lastIndex][1], current[1])
            }
        }
        
        return merged.toTypedArray()
    }
}
